Implement some more slice allocation functions

TethysSvensson opened the pull request:
This implements the feature requested in #41.

ztlpn commented on src/lib.rs:
````diff
@@ -595,6 +615,114 @@ impl Bump {
+    /// ```
+    #[inline(always)]
+    #[allow(clippy::mut_from_ref)]
+    pub fn alloc_slice_fill_copy<T: Copy>(&self, len: usize, value: T) -> &mut [T] {
````
Hey @TethysSvensson thanks for implementing the feature I requested, that was quick! One remark: I messed up the order of arguments in the `vec![]` macro in my request (as I always do, it seems impossible to remember!) - the correct order is `vec![value, len]`. So I think here we should use the same order fo consistency.

TethysSvensson replied:
You are right, that makes more sense.

TethysSvensson commented:
> Implementation looks great -- can you also add some tests for these new methods? Thanks!

Do you mean beyond the ones in the doctests? Which tests would you like me to implement?

fitzgen commented:
> Do you mean beyond the ones in the doctests? Which tests would you like me to implement?

* tests for the `len == 0` fast path

* tests for when the layout overflows in rounding up and in the multiply

* would be good to get some quickchecks in `tests/quickchecks.rs` as well:

```rust
quickcheck! {
    // ...

    fn alloc_slices(allocs: Vec<(u8, usize)>) -> () {
        let b = Bump::new();
        let mut allocated = vec![];
        for (val, len) in allocs {
            let s = b.alloc_slice_copy(val, len); // maybe need to place a max on `len`.

            // todo: assert that the slice is all `val`

            let range = (s.as_ptr(), s.as_ptr().add(s.len()));
            for r in &allocated {
                assert!(!overlaps(*r, range));
            }
            allocated.push(range);
        }
    }
}
```

Thanks!

TethysSvensson commented:
I think we should also provide a function that takes a closure and calls it for every element. That would also be useful internally to avoid code duplication.

fitzgen commented:
> I think we should also provide a function that takes a closure and calls it for every element. That would also be useful internally to avoid code duplication.

:+1:

TethysSvensson commented:
I have rebased this on top of master and added a new commit.

TethysSvensson commented:
I am still missing more tests before this can merge.

TethysSvensson commented:
@fitzgen I've implemented the tests now. Please take a look.

I also moved the special-casing of ZSTs into the core allocation instead of dealing with it in `alloc_slice`. There is a test that the allocation path works as expected. I am not sure how to test the deallocation/reallocation path. If you have any ideas, I'll be happy to add tests, but otherwise I think the code is simple enough that it works as is.
